v3/src/models/formula/functions/string-functions.ts

```diff
@@ -0,0 +1,210 @@
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const text = valueToString(args[0]), index = Number(args[1])
+      return text.charAt(index - 1) // 1-based index for CODAP string functions
+    }
+  },
+  // combine(text) Returns the concatenation of text all the values of the attribute. This is an aggregate function.
+  combine: {
+    numOfRequiredArguments: 1,
+    evaluate: (...args: FValue[]) => {
+      // todo: Implement this function
+      return "NYI"
+    }
+  },
+  // concat(string1, string2, …) Returns a string that is the concatenation of the string arguments.
+  concat: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      return args.map(valueToString).join("")
+    }
+  },
+  // endsWith(text, suffix) Returns true if text ends with suffix, otherwise false.
+  endsWith: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const text = valueToString(args[0]), suffix = valueToString(args[1])
+      if (!suffix) return true
+      if (!text) return false
+      return text.endsWith(suffix)
+    }
+  },
+  // findString(stringToLookIn, stringToFind, start) Takes three arguments, returning the position of
+  // stringToFind in stringToFind starting from start. The first character of stringToLookIn is numbered 1.
+  // Returns 0 if stringToFind is not found. The third argument need not be present and defaults to 1.
+  findString: {
+    numOfRequiredArguments: 3,
+    evaluate: (...args: FValue[]) => {
+      const stringToLookIn = valueToString(args[0]),
+        stringToFind = valueToString(args[1]),
+        start = Number(args[2]) || 1
+      return stringToLookIn.indexOf(stringToFind, start - 1) + 1
+    }
+  },
+  // includes(stringToLookIn, stringToFind) Takes two arguments and returns true if the second argument
+  // is a substring of the first (also treated as a string).
+  includes: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const stringToLookIn = valueToString(args[0]),
+        stringToFind = valueToString(args[1])
+      return stringToLookIn.includes(stringToFind)
+    }
+  },
+  // join(delimiter, string1, string2, …) Returns the string formed by concatenating its string arguments, each
+  // seperated by its delimiter argument.
+  join: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const delimiter = valueToString(args[0])
+      return args.slice(1).map(valueToString).join(delimiter)
+    }
+  },
+  // patternMatches(stringToLookIn, pattern) Returns the number of times in string_to_look_in the string or
+  // regular expression represented by pattern occurs. The search is case-insensitive.
+  patternMatches: {
+    numOfRequiredArguments: 2,
+    evaluate: (...args: FValue[]) => {
+      const stringToLookIn = valueToString(args[0]),
+        pattern = valueToString(args[1])
+      return (stringToLookIn.match(new RegExp(pattern, "gmi")) || []).length
```

Comment:
Strings, particularly user-entered strings, should always be escaped before using them as regular expressions.
```suggestion
      return (stringToLookIn.match(new RegExp(escapeStringRegexp(pattern), "gmi")) || []).length
```